Revert BackwardCompatibilityInterceptor's behavior of changing /apis to /oapi URLs

rohanKanojia opened the pull request:
Revert #2373

This commit reverts 293ab9d445291de389d438f4a857928722230dba which was added in order
to fix #2373. This behavior was causing problems with Strimzi's Kubernetes Client
upgrade: https://github.com/strimzi/strimzi-kafka-operator/pull/3553 . On bisecting
the failing test, I found this commit as culprit. Git bisect log:
```
kubernetes-client : $ git bisect log
git bisect start
# bad: [ea09419284b32e6da70aadfaf9022485f2bec594] Merge pull request #2478 from rohanKanojia/pr/migrate-github-actions-java8
git bisect bad ea09419284b32e6da70aadfaf9022485f2bec594
# good: [c5fe5488dc15ed3d4795a3b3d676549c24c2e744] Post v4.10.3 release chores
git bisect good c5fe5488dc15ed3d4795a3b3d676549c24c2e744
# bad: [262e35f09be835bf5f3b5fbc4c16791224fbc390] Merge pull request #2381 from rohanKanojia/pr/issue2308
git bisect bad 262e35f09be835bf5f3b5fbc4c16791224fbc390
# bad: [d13a86480f699735dcc8e0ece90c24d4e49079e8] Merge pull request #2393 from rohanKanojia/pr/fix-failing-resource-it
git bisect bad d13a86480f699735dcc8e0ece90c24d4e49079e8
# good: [f1d695d59ce50a9067639df0b79855f03096adc8] Merge pull request #2371 from manusa/fix/mockserver-post
git bisect good f1d695d59ce50a9067639df0b79855f03096adc8
# good: [0b1e0668788aaf2d7f0e706f111ca67456cbb36c] Merge pull request #2361 from fabric8io/dependabot/maven/org.jboss-jandex-2.2.1.Final
git bisect good 0b1e0668788aaf2d7f0e706f111ca67456cbb36c
# bad: [14cae945ee20ce1ac980d492e53d388aa93eb371] Merge pull request #2372 from rohanKanojia/pr/issue2292
git bisect bad 14cae945ee20ce1ac980d492e53d388aa93eb371
# bad: [f6d4e8fedd2d7c36a5608e5f3a4fd73070e51559] Merge pull request #2375 from rohanKanojia/pr/issue2373
git bisect bad f6d4e8fedd2d7c36a5608e5f3a4fd73070e51559
# bad: [293ab9d445291de389d438f4a857928722230dba] Fix #2373: Unable to create a Template on OCP3
git bisect bad 293ab9d445291de389d438f4a857928722230dba
# first bad commit: [293ab9d445291de389d438f4a857928722230dba] Fix #2373: Unable to create a Template on OCP3
```

Handling OpenShift old /oapi requests seem to be handled in OpenShiftOperation where we check `config.isOpenshiftApiGroupsEnabled()` to modify OperationContext.

#2373 can be fixed by adding apiVersion in OpenShiftOperation.

## Description
<!--
Thank a lot for taking time to contribute to Fabric8 <3!

Please provide a description of what your PR does providing a link (if applicable) to the issue it fixes. It is
really helpful for people who would review your code.
-->

## Type of change
<!---
What types of changes does your code introduce? Put an `x` in all the boxes that apply
-->
 - [X] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [X] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [ ] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [ ] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [X] I tested my code in OpenShift

<!--
Integration tests (https://github.com/fabric8io/kubernetes-client/tree/master/kubernetes-itests)
Please check integration tests and provide/improve tests if applicable.

Open your PR in Draft mode and verify all of the applicable Checklist items before marking your pull request as ready for review
-->

manusa commented:
You're "reverting" too much, some of the fixes provided after (like the one for createOrReplace #2467) are reverted too.

rohanKanojia commented:
I reverted BackwardCompatibilityInterceptor to it's 4.10.3 state. I think #2467 was also caused by my change in #2375 . I haven't reverted tests for #2467 I added(they're passing with this change):

https://github.com/fabric8io/kubernetes-client/blob/323be1cbfe2b3a2b5db97625337884123f32f4d8/kubernetes-itests/src/test/java/io/fabric8/openshift/DeploymentConfigIT.java#L93-L97

https://github.com/fabric8io/kubernetes-client/blob/323be1cbfe2b3a2b5db97625337884123f32f4d8/kubernetes-itests/src/test/java/io/fabric8/openshift/RouteIT.java#L99-L107

manusa commented:
> I reverted BackwardCompatibilityInterceptor to it's 4.10.3 state. I think #2467 was also caused by my change in #2375 . 

I'm not so sure about that.

The thing is that with the "new" `createOrReplace` procedure, we are relying on a `409 Conflict` HTTP status code response to trigger the replace request.

If you don't keep this check in the interceptor:
```java
 if (!response.isSuccessful() && response.code() == HttpURLConnection.HTTP_NOT_FOUND)
```
The interceptor will "kick in" whenever a resource needs to be replaced (too).

Regarding tests not failing, I'm guessing the tests you implemented in #2467 fix did not include an OpenShift resource specific replace case scenario.

rohanKanojia commented:
I think you're right. Right now BackwardsCompatibiityInterceptor seems to be only converting `/oapi`  URLs to `/apis`. Client sends `/apis` URLs for both OpenShift 3.10.0 and 3.11.0, this is the reason tests are passing on both. I should add that 404 check.

rohanKanojia commented:
[merge]

tombentley commented:
@rohanKanojia please could you tell me when there might be a release with this change in?

manusa commented:
Hi @tombentley 
If there are no impediments, we'll be cutting a release today.

tombentley commented:
Awesome, thanks!

manusa commented:
Hi @tombentley 
In case you missed it, new release is available https://github.com/fabric8io/kubernetes-client/releases/tag/v4.12.0 on Maven Central.
